### PR TITLE
Create mogoose user stats prototype

### PR DESCRIPTION
**Pull request by aleklisi**
This PR creates a module reporting MongooseIM usage data to Google Analytics.

Currently reporting:
 - Number of configured hosts
 - Open source mods with backends (if configured)

### CODE REVIEW

**Comment by mongoose-im** [comment]
[7418.1](https://travis-ci.org/esl/MongooseIM/jobs/605404580) / Erlang 22.0 / small_tests / ac586cbd76ed4a47767d1cb3f90613a18a53d292
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7418/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7418/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-10-31_09.29.24/lib.mongooseim.logs/run.2019-10-31_09.29.24/suite.log.html)

File: src/mongoose_user_stats.erl [hunk 1]
```diff
@@ -0,0 +1,76 @@
+-module(mongoose_user_stats).
+
+-author('aleksander.lisiecki@erlang-solutions.com').
+
+% TODO replace tid to official ESL one
+-define(URL_BASE, "https://www.google-analytics.com/collect?v=1&tid=UA-151110014-1&t=event").
+
+-export([report_user_stats/0]).
```
Comment (michalwski) [inline src/mongoose_user_stats.erl]: I think the function could be named just `report`. The full call `mongoose_user_stats:report` clearly identifies that we are reporting user stats (at least to me).

File: src/mongoose_user_stats.erl [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+report_user_stats() ->
+    IsAllowed = ejabberd_config:get_local_option_or_default(is_report_user_stats_allowed, true),
```
Comment (michalwski) [inline src/mongoose_user_stats.erl]: By default reporting should be `disabled`, we change it to `enabled` when we think this functionality is ready.
Also I'm wondering if name like `report_user_stats` with possible values `enabled`, `disabled` would be better here. What do you think?

File: src/mongoose_user_stats.erl [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+% Functions are spawned and not linked, as MongooseIM should not care if they fail or not.
+% Moreover the MongooseIM's start should not be blocked.
+report_user_stats(true) ->
+    % Data used for more then one report
+    Hosts = ejabberd_config:get_global_option(hosts),
+    Reports = [
+        fun() -> report_number_of_hosts(Hosts) end,
+        fun() -> report_used_modules(Hosts) end
+    ],
+    lists:foreach(
```
Comment (michalwski) [inline src/mongoose_user_stats.erl]: Do you think the following
```erlang
[spawn(Fun) || Fun <- Reports]
```

is more readable than the `lists:foreach`?

File: src/mongoose_user_stats.erl [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    report(EventCategory, EventAction, empty).
+
+report(EventCategory, EventAction, EventLabel) ->
+    MaybeLabel = case EventLabel of
+        empty ->
+            [];
+        AnyEventLabel ->
+            LstEventLabel = term_to_string(AnyEventLabel),
+            ["&el=", LstEventLabel]
+    end,
```
Comment (michalwski) [inline src/mongoose_user_stats.erl]: Could you extract this `case` statement to a function?

File: src/mongoose_user_stats.erl [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    URL = string:join(ListUrl, ""),
+    lager:debug("~p reported = ~p", [?MODULE, URL]),
+    httpc:request(URL).
+
+term_to_string(Term) ->
+    R= io_lib:format("~p",[Term]),
+    lists:flatten(R).
+
+client_id() ->
+    % TODO in the later implementation store client's ID in eg mnesia table and report stats with the same ID for the same client
```
Comment (michalwski) [inline src/mongoose_user_stats.erl]: The `global` config is a good candidate to store this ID.

File: src/mongoose_user_stats.erl [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    LstEventAction = term_to_string(EventAction),
+    ListUrl = [
+        ?URL_BASE,
+        "&cid=", LstClientId,
+        "&ec=", LstEventCategory,
+        "&ea=", LstEventAction
+        ] ++ MaybeLabel,
+    URL = string:join(ListUrl, ""),
+    lager:debug("~p reported = ~p", [?MODULE, URL]),
+    httpc:request(URL).
```
Comment (michalwski) [inline src/mongoose_user_stats.erl]: I'm not fun of sending a request for every report. There is still some time, so I think we can introduce some batching and send less but bigger requests.

**Comment by mongoose-im** [comment]
[7425.1](https://travis-ci.org/esl/MongooseIM/jobs/607670711) / Erlang 22.0 / small_tests / b21e4f3a2dae3caa53547196785e8c5919e1d3e1
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7425/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7425/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-11-05_14.36.55/lib.mongooseim.logs/run.2019-11-05_14.36.55/suite.log.html)

---

[7425.2](https://travis-ci.org/esl/MongooseIM/jobs/607670712) / Erlang 22.0 / internal_mnesia / b21e4f3a2dae3caa53547196785e8c5919e1d3e1
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7425/internal_mnesia.22.0)/ [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7425/internal_mnesia.22.0/big/ct_run.test%40travis-job-626fa9d2-3e67-406d-aae4-2bae95bf2402.2019-11-05_14.37.17/index.html)
**OK: 1273** / Failed: 0 / **User-skipped: 138** / Auto-skipped: 0 

---

[7425.3](https://travis-ci.org/esl/MongooseIM/jobs/607670713) / Erlang 22.0 / odbc_mssql_mnesia / b21e4f3a2dae3caa53547196785e8c5919e1d3e1
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7425/odbc_mssql_mnesia.22.0)/ [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7425/odbc_mssql_mnesia.22.0/big/ct_run.test%40travis-job-a8d57e6d-ede4-449e-884e-d61f6ebc7dc6.2019-11-05_14.38.15/index.html)
**OK: 2492** / Failed: 0 / **User-skipped: 199** / Auto-skipped: 0 

---

[7425.4](https://travis-ci.org/esl/MongooseIM/jobs/607670714) / Erlang 22.0 / mysql_redis / b21e4f3a2dae3caa53547196785e8c5919e1d3e1
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7425/mysql_redis.22.0)/ [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7425/mysql_redis.22.0/big/ct_run.test%40travis-job-8e64f8f8-f829-44a1-823e-4d7afc89ce53.2019-11-05_14.51.07/index.html)
**OK: 2487** / Failed: 0 / **User-skipped: 204** / Auto-skipped: 0 

---

[7425.7](https://travis-ci.org/esl/MongooseIM/jobs/607670722) / Erlang 22.0 / elasticsearch_and_cassandra_mnesia / b21e4f3a2dae3caa53547196785e8c5919e1d3e1
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7425/elasticsearch_and_cassandra_mnesia.22.0)/ [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7425/elasticsearch_and_cassandra_mnesia.22.0/big/ct_run.test%40travis-job-5e2ab68e-c7a4-4777-8b74-13c12846bc18.2019-11-05_15.17.21/index.html)
**OK: 328** / Failed: 0 / **User-skipped: 28** / Auto-skipped: 0 

---

[7425.6](https://travis-ci.org/esl/MongooseIM/jobs/607670721) / Erlang 22.0 / ldap_mnesia / b21e4f3a2dae3caa53547196785e8c5919e1d3e1
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7425/ldap_mnesia.22.0)/ [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7425/ldap_mnesia.22.0/big/ct_run.test%40travis-job-06553ec2-4072-45c1-8634-b66b127d65ad.2019-11-05_15.07.25/index.html)
**OK: 1242** / Failed: 0 / **User-skipped: 169** / Auto-skipped: 0 

---

[7425.5](https://travis-ci.org/esl/MongooseIM/jobs/607670717) / Erlang 22.0 / riak_mnesia / b21e4f3a2dae3caa53547196785e8c5919e1d3e1
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7425/riak_mnesia.22.0)/ [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7425/riak_mnesia.22.0/big/ct_run.test%40travis-job-72c586d2-a12f-4254-8588-337b95822f24.2019-11-05_15.04.15/index.html)
**OK: 1406** / Failed: 0 / **User-skipped: 145** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[7426.1](https://travis-ci.org/esl/MongooseIM/jobs/607692156) / Erlang 22.0 / small_tests / cb75b825538305447edb6b78e0f7f31c205c3506
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7426/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7426/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-11-05_15.42.54/lib.mongooseim.logs/run.2019-11-05_15.42.55/suite.log.html)

---

[7426.2](https://travis-ci.org/esl/MongooseIM/jobs/607692157) / Erlang 22.0 / internal_mnesia / cb75b825538305447edb6b78e0f7f31c205c3506
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7426/internal_mnesia.22.0)/ [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7426/internal_mnesia.22.0/big/ct_run.test%40travis-job-f2df4938-6774-4566-9935-cef72141c0bf.2019-11-05_15.43.45/index.html)
**OK: 1273** / Failed: 0 / **User-skipped: 138** / Auto-skipped: 0 

---

[7426.3](https://travis-ci.org/esl/MongooseIM/jobs/607692158) / Erlang 22.0 / odbc_mssql_mnesia / cb75b825538305447edb6b78e0f7f31c205c3506
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7426/odbc_mssql_mnesia.22.0)/ [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7426/odbc_mssql_mnesia.22.0/big/ct_run.test%40travis-job-04d7bd05-0e1a-447a-b27e-bc73c2be7313.2019-11-05_15.56.58/index.html)
**OK: 2492** / Failed: 0 / **User-skipped: 199** / Auto-skipped: 0 

---

[7426.4](https://travis-ci.org/esl/MongooseIM/jobs/607692159) / Erlang 22.0 / mysql_redis / cb75b825538305447edb6b78e0f7f31c205c3506
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7426/mysql_redis.22.0)/ [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7426/mysql_redis.22.0/big/ct_run.test%40travis-job-8c46ceb9-f4af-4c13-80e3-6c3a2a560f34.2019-11-05_16.07.40/index.html)
**OK: 2487** / Failed: 0 / **User-skipped: 204** / Auto-skipped: 0

File: src/mongoose_user_stats.erl (hunk 6 above)
Comment (aleklisi) [reply src/mongoose_user_stats.erl]: I added batching :)

**Comment by mongoose-im** [comment]
[7427.1](https://travis-ci.org/esl/MongooseIM/jobs/608086927) / Erlang 22.0 / small_tests / 8105f354e19114164868e5ed73df57ac50654be1
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7427/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7427/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-11-06_09.17.02/lib.mongooseim.logs/run.2019-11-06_09.17.03/suite.log.html)

File: src/mongoose_user_stats.erl (hunk 2 above)
Comment (michalwski) [reply src/mongoose_user_stats.erl]: Yes, finally I want it to be enabled by default. Currently this is not ready to be used by our community, that's why I want it to be disabled by default. When we decide it's ready to use in real world, we'll change to enabled by default.

File: src/mongoose_user_stats.erl (hunk 5 above)
Comment (michalwski) [reply src/mongoose_user_stats.erl]: Correct, we want it to persistent.

File: big_tests/tests/mongoose_user_stats_SUITE.erl [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+end_per_testcase(_CaseName, Config) ->
+    Config.
+
+%%--------------------------------------------------------------------
+%% Tests
+%%--------------------------------------------------------------------
+
+
+number_of_hosts_is_reported_to_google_analytics_when_mim_starts(_Config) ->
```
Comment (michalwski) [inline big_tests/tests/mongoose_user_stats_SUITE.erl]: I think the name of the test case doesn't reflect that it also check if modules are reported correctly.

File: src/mongoose_user_stats.erl [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+% https://developers.google.com/analytics/devguides/collection/protocol/v1/devguide#batch-limitations
+% A maximum of 20 hits can be specified per request.
+send_reports(ReportUrl, Lines) when length(Lines) =< 20 ->
+    Headers = [],
+    ContentType = "",
+    Body = string:join(Lines, "\n"),
+    Request = {ReportUrl, Headers, ContentType, Body},
+    Res = httpc:request(post, Request, [], []),
+    lager:error("Res = ~p", [Res]);
```
Comment (michalwski) [inline src/mongoose_user_stats.erl]: A leftover?

**Comment by mongoose-im** [comment]
[7428.1](https://travis-ci.org/esl/MongooseIM/jobs/608116102) / Erlang 22.0 / small_tests / cc4efe8df2f42616dcba15dd0e96b8c772307180
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7428/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7428/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-11-06_10.39.25/lib.mongooseim.logs/run.2019-11-06_10.39.26/suite.log.html)

File: src/mongoose_user_stats.erl (hunk 8 above)
Comment (aleklisi) [reply src/mongoose_user_stats.erl]: yes

File: big_tests/tests/mongoose_user_stats_SUITE.erl (hunk 7 above)
Comment (aleklisi) [reply big_tests/tests/mongoose_user_stats_SUITE.erl]: I will rename it

**Comment by mongoose-im** [comment]
[7429.1](https://travis-ci.org/esl/MongooseIM/jobs/608165168) / Erlang 22.0 / small_tests / 8b3202001cbc970fa0f85c8295426c95ffb1f4c2
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2523/7429/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2523/7429/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-11-06_12.36.14/lib.mongooseim.logs/run.2019-11-06_12.36.15/suite.log.html)

**Pull request merged**